otter/convergence/gathering.py

```diff
@@ -111,8 +114,49 @@ def group_id(s):
                             filter(server_predicate),
                             filter(has_group_id))
 
-    eff = get_all_server_details(changes_since)
-    return eff.on(servers_apply)
+    return get_all_server_details(changes_since).on(servers_apply)
+
+
+def merge_servers(first, second):
+    """
+    Return servers got by merging the two sets of servers where second
+    takes precedence over first
+
+    :param list first: List of server dicts
+    :param list second: List of server dicts
+    :return: List of merged server dicts
+    """
+    def sdict(servers):
+        return {s['id']: s for s in servers}
+
+    return merge(sdict(first), sdict(second)).itervalues()
+
+
+@do
+def get_scaling_group_servers(tenant_id, group_id, now,
+                              cache_class=CassScalingGroupServersCache):
+    """
+    Get a group's servers taken from cache if it exists. Updates cache
+    if it is empty from newly fetched servers
+    # NOTE: This function takes tenant_id even though the whole effect is
+    # scoped on the tenant because cache calls require tenant_id. Should
+    # they also not take tenant_id and work on the scope?
+
+    :return: Servers as iterator of dicts
+    :rtype: Effect
+    """
+    coll = cache_class(tenant_id, group_id)
+    cache, last_update = yield coll.get_servers(tenant_id, group_id)
+    if last_update is None or last_update - now >= timedelta(days=30):
+        last_update = now - timedelta(days=30)
+        changes = (yield get_all_scaling_group_servers(last_update))[group_id]
+        current = (yield get_all_server_details())[group_id]
```

Comment:
Should this be `(yield get_all_scaling_group_servers())[group_id]`?  `get_all_server_details` doesn't group them by group ID I think.

Non-blocking idle thought - I'm also wondering if we should rename these functions?   (the `get_all_scaling_group_servers` change was good, I'm just wondering if we should name the ones that gather from nova explicitly as gathering from nova, as opposed to this function, which does both cache checking and gathering.